### ISSUE TITLE
Relationshop with pythonanywhere fork

### ISSUE THREAD

**Issue by jasontibbitts** [shellinabox/shellinabox#334]
Sorry for opening an issue but I'm not sure where this can be discussed.

I was looking at the shellinabox package in Fedora and found that it's using the pythonanywhere fork which hasn't been maintained in some time.  I was wondering how the two forks relate to each other and whether it might be reasonable to simply switch the Fedora package over to this fork.  I haven't yet dug into it to see if there's any significant differences in functionality between the two, but I can always make a separate package if necessary.

The Fedora ticket I've opened for this is at https://bugzilla.redhat.com/show_bug.cgi?id=1252109

Thanks for your input.

**Comment by a-detiste** [comment]
Hi,

@KLuka already took care of reviewing all the various forks floating around including this one + imported & triaged all the issues created at original GoogleCode repository.

I helped a bit with the Debian package as I use this tool everyday on RaspBian, which is 99,9% Debian testing.

There were forks doing a bit of dumb things, like patching generated build files, but not the source that it came from. @Kluka started with a clean fork. Now all auto-generated cruft & subversion stuff has been expunged, this can't happen anymore (but your distro will need to call autoreconf first)

---

https://apps.fedoraproject.org/packages/shellinabox/contents/
 
I see there that Fedora is shipping a systemd service; it would be interresting to have one shipped upstream as a reference. (I haven't used RedHat since 5.2 in 1999, can't find where I can read the content of this file)

**Comment by jasontibbitts** [comment]
Awesome, thanks.

You can see the repository for the Fedora package, including the systemd unit, at http://pkgs.fedoraproject.org/cgit/shellinabox.git/tree/

Fedora is carrying one patch to add -oProxyCommand=none to the ssh command line.  I'm not sure where this patch came from or why it is needed; there's no provenance information in the patch or spec.

I have an updated and cleaned package building locally but before doing anything I'll wait to see what the current Fedora package maintainer would like to do.  I can send the spec along once I'm done tweaking if you'd like to see it.

**Comment by jasontibbitts** [comment]
The patch comes from the following bug report:
https://bugzilla.redhat.com/show_bug.cgi?id=1013974

It basically forces ProxyCommand off in case sshd_config has it set (which happens in some setups using IPA).

**Comment by a-detiste** [comment]
Duh, what is IPA ? This is the only reference I could find:

https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Identity_Management_Guide/host-keys.html

**Comment by jasontibbitts** [comment]
It's sort of like an active directory thing.  https://www.freeipa.org/page/Main_Page

**Comment by a-detiste** [comment]
about the systemd service:

`EnvironmentFile=-/etc/sysconfig/shellinaboxd` with a `-`means  that /etc/sysconfig/shellinaboxd is optional from systemd point of view; but if this file is missing, ExecStart evaluates to `/usr/sbin/shellinaboxd -u -g --cert= --port=` and service fail.

There is now a trend in Debian to get rid of /etc/default (same thing as /etc/sysconfig); by moving these configuration bits in the systemd service; that can be overloaded if needed with `systemctl edit --full shellinabox`. Then users benefits of standard tools like systemd-delta to see what was tweaked....

So you might want to do the same or not, but a upstream-provided systemd service should really not depend on /etc/default or /etc/sysconfig

**Comment by jasontibbitts** [comment]
I agree; I don't make any guarantees that the systemd unit is remotely correct as I've only started looking at this package today.  Fedora has long been trying to eliminate /etc/sysconfig files (as we've been doing the systemd thing for many years now) but this package never caught up and I suspect someone's afraid of breaking things.  I'll clean it up in my package but I haven't done much more than get it building.

**Comment by KLuka** [comment]
Hello guys, it would be great if this fork can be used in Fedora distributions since a lot of people are using this package on CentOS... And as @a-detiste said this fork has a lot of fixes included and even some new features, we are also trying to keep it stable as possible :smile: 

About the `ProxyCommand` I don't see any problems if this patch is pushed to upstream master. Probably other users also have issues, if this option is set in global config...

For the systemd config I also have no objection, if you guys crate some working example files. I would be glad to pull them in :wink:

**Comment by jasontibbitts** [comment]
My goal is to get this fork into Fedora somehow.  Right now it mostly depends on the current maintainer of the "shellinabox" package, but there's no reason I can't get you a clean (and tested) unit file and a rebased version of the ProxyCommand patch immediately.  Unfortunately I'm going to be on a plane soon (heading to a Fedora conference) but I should be able to give you something in the next few days.  That will also give the current maintainer a chance to respond.

**Comment by KLuka** [comment]
Ok great and there is no rush ;)

**Comment by KLuka** [comment]
Since this RPM is also used on CentOS I would also like to tell you that there are some problems on CentOS 7 and possibly on other newer Fedora based distros. Problem was reported in #327 and it is 100% reproducible.

On CentOS 7 shellinabox cannot properly execute `/bin/login` and we get `login -- <user>: no shell: Permission denied ` error message displayed in browser...

I suspect that SELinux is involved in this issue. Problem is probably related to the fact that on newer distros shellineabox process runs as `unconfined_service_t`.  I don't know how can we fix this? Do you have any experience with SELinux policy?

Here is the diff from `ps axZ` command on both systems:

```diff
-system_u:system_r:initrc_t:s0    4595 ?        Ss     0:00 shellinaboxd -u shellinabox -g shellinabox --cert=/var/lib/shellinabox --port=4200 --background=/var/run/shellinaboxd.pid --user-css Normal:+black-on-white.css,Reverse:-white-on-black.css --disable-ssl-menu -s /:LOGIN
+system_u:system_r:unconfined_service_t:s0 2346 ? Ss   0:00 shellinaboxd -u shellinabox -g shellinabox --cert=/var/lib/shellinabox --port=4200 --disable-ssl-menu -s /:LOGIN
```

Also since CentOS 7 `shellinaboxd.service` file is used by default (?), so user CSS options are missing in ExecService definition...

**Comment by scaronni** [comment]
Hello, I'm the maintainer of shellinabox in Fedora/EPEL, I've been pointed to this in a bug report. Frankly I did not know this fork existed, otherwise I would have switched to it much earlier.

Right now I'm mostly using Guacamole [1] [2], but I'm still maintaining shellinabox and would like to update it.

I've made a few test builds in Koji for all releases, can you guys please tell me if it works for your environment? Right now I have two issues:

- I can not make it work with --disable-ssl, --disable-ssl-menu or -t. It seems --disable-ssl-menu is ignored and the SSL redirect happens anyway.
- I can not rebuild it on RHEL/CentOS 5.

Regarding the environment file (/etc/sysconfig/shellinaboxd), the package was conceived before systemd appeared and I am rebuilding the same package (it has conditionals inside) for all the releases. Since I can not change the behaviour mid-release I will remove the environment file in Fedora 23+. Can anyone make some test on CentOS/RHEL/Fedora if I prepare the builds?

[1] http://guac-dev.org/
[2] http://negativo17.org/guacamole-on-fedora-and-centosrhel-6/

**Comment by scaronni** [comment]
I've opened a bug for the CentOS/RHEL 5 part on autotools: #336

Builds are here, if people want to test (environment file still there):

https://koji.fedoraproject.org/koji/packageinfo?packageID=13996

**Comment by KLuka** [comment]
Hello @scaronni and thanks for stopping by :+1: 

It would be great, if shellinabox package can be updated for Fedora based distros. But I agree that we should solve the issues first :)

> I can not make it work with --disable-ssl, --disable-ssl-menu or -t. It seems --disable-ssl-menu is ignored and the SSL redirect happens anyway.

Hm, that is strange. I tested your RPM `shellinabox-2.17-2.el7.x86_64.rpm` just now on `CentOS Linux release 7.1.1503 (Core)` and it seams to work fine...

How did you change your configuration? You can check the actual config of started daemon with `ps`. You can also try to run it manually and check/paste the debug output. Like this for example: 
```
/usr/sbin/shellinaboxd --debug --disable-ssl
```
> On CentOS 7 shellinabox cannot properly execute /bin/login and we get login -- <user>: no shell: Permission denied error message displayed in browser...

Also your package seams to solve this issue although shellinabox is running with LOGIN service. IDK why is this working now?

> I can not rebuild it on RHEL/CentOS 5.

Handling it in #336 :smile:

**Comment by scaronni** [comment]
> Hm, that is strange. I tested your RPM shellinabox-2.17-2.el7.x86_64.rpm just now on CentOS Linux release 7.1.1503 (Core) and it seams to work fine...

Yes, nevermind, was testing it on a dirty system I use for everything.

> Sorry I forgot to reenable SELinux. Now I get the same message again :/ I would really appreciate some help on this one.

Looking into it. Last time I used it in production I was using shellinabox as an SSH jump host for other systems.

> Handling it in #336 

Fixed, thanks. Rebuilt in the same place: https://koji.fedoraproject.org/koji/packageinfo?packageID=13996

Will defer the update in epel 6/7 and Fedora 21+ after solving the SELinux issue. Thanks!

**Comment by scaronni** [comment]
Well, sorry, I changed my mind. I was looking at the detailed changelog from v2.14 to v2.17 and the upgrade is absolutely needed, I will release an update immediately and then proceed with integrating the SELinux policy upstream.

Thanks.

**Comment by KLuka** [comment]
I recommend that you take the latest version from master and I will tag it as 2.18. Is this OK?

**Comment by scaronni** [comment]
Sure, no problem. Thanks.

**Comment by KLuka** [comment]
Ok, I tagged [release 2.18](https://github.com/shellinabox/shellinabox/releases/tag/v2.18). I recommend that this release is used for Fedora RPM package update, since it contains all the latest improvements and fixes...

For SELinux problem on CentOS 7 I opened another issue (#342), where we can discuss and hopefully solve this thing...

**Comment by uudruid74** [comment]
Hey - I have my own shellinabox fork ... https://github.com/uudruid74/shellinabox_fork

It includes most of what I could find and also quite a few bug fixes of my own, particularly in the color handling which should make things a bit nicer for tmux users.   I'm not even sure where to begin to merge these, but I think it would be a good idea to get our trees merged.   Mine was forked from the pythonanywhere source as well, so there should be some way to do it sanely?

**Comment by uudruid74** [comment]
One other note.  Is the pre-processor really necessary?   Is it serving any purpose anymore?  I've been updating only the vt100.js rather than the jspp file since its a lot easier to deal with.   Any chance we can ditch the jspp file and just keep the js in git?
